rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -0,0 +1,273 @@
+package coop.rchain.rholang.interpreter
+
+import cats._
+import cats.data._
+import cats.implicits._
+
+import scala.annotation.tailrec
+import scala.collection.immutable.Stream
+
+object SpatialMatcher {
+  type FreeMap            = Map[Int, Par]
+  type OptionalFreeMap[A] = StateT[Option, FreeMap, A]
+  type NonDetFreeMap[A]   = StateT[Stream, FreeMap, A]
+
+  def possiblyRemove[T](needle: T, haystack: List[T]): Option[List[T]] = {
+    val (before, after) = haystack.span(x => x != needle)
+    after match {
+      case Nil       => None
+      case _ :: tail => Some(before ++ tail)
```

Comment:
This won't work properly if there are multiple occurences of the `needle` in the `haystack` . Why not `haystack.filterNot(_ == needle)` ?

Reply:
@goral09 The point of this function is to remove it once. If I wanted to remove it every time, I would use something like what you listed inline.